In [1]:
!pip install matplotlib
!pip install numpy
!pip install pandas
!pip install statsmodels
!pip install seaborn
!pip install itertools

ERROR: Could not find a version that satisfies the requirement itertools (from versions: none)
ERROR: No matching distribution found for itertools


In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import statsmodels.api as sm
import seaborn as sns
import itertools

sns.set(style="dark")
plt.style.use("ggplot")
%matplotlib inline

# Load in datasets & clean

In [6]:
trash_amount = pd.read_csv('Data/AMOUNT OF TRASH (SIPSN).csv')
trash_amount.drop(columns=['Unnamed: 5','https://sipsn.menlhk.go.id/sipsn/public/data/timbulan'], inplace=True)
trash_amount['Per day (ton)'] = trash_amount['Per day (ton)'].str.replace(',','') # some numbers have , 
trash_amount['Per year (ton)'] = trash_amount['Per year (ton)'].str.replace(',', '') # some numbers have , 
trash_amount

Tahun        Provinsi           Kabupaten/Kota Per day (ton)  \
0      2019            Aceh        Kab. Aceh Selatan         95.93   
1      2019            Aceh         Kab. Aceh Tengah        100.97   
2      2019            Aceh            Kab. Simeulue         43.96   
3      2019            Aceh           Kab. Gayo Lues         37.64   
4      2019            Aceh        Kab. Bener Meriah         64.89   
...     ...             ...                      ...           ...   
1029   2022  Sumatera Utara  Kab. Humbang Hasundutan         59.33   
1030   2022  Sumatera Utara               Kota Medan       1722.60   
1031   2022  Sumatera Utara    Kota Pematang Siantar        241.43   
1032   2022  Sumatera Utara       Kota Tebing Tinggi        124.63   
1033   2022  Sumatera Utara     Kota Padangsidimpuan        161.82   

     Per year (ton)  
0          35013.28  
1          36853.74  
2          16044.14  
3          13738.60  
4          23683.10  
...             ...  
1029       21653.73  
1030      628749.22  
1031       88121.44  
1032       45488.20  
1033       59065.98  

[1034 rows x 5 columns]

In [7]:
trash_comp = pd.read_csv('Data/TRASH COMPOSITION (SIPSN) .csv')
trash_comp.drop(columns=['Unnamed: 12',	'https://sipsn.menlhk.go.id/sipsn/public/data/komposisi'], inplace=True)
trash_comp

Year        Province                     Town  Food (%)  Wood (%)  \
0     2022            Aceh        Kab. Aceh Selatan       NaN       NaN   
1     2022            Aceh       Kab. Aceh Tenggara      41.0      11.0   
2     2022            Aceh          Kab. Aceh Timur      37.0      14.0   
3     2022            Aceh          Kab. Aceh Barat      41.8      13.3   
4     2022            Aceh               Kab. Pidie       8.5       7.5   
...    ...             ...                      ...       ...       ...   
1029  2019  Sumatera Utara  Kab. Humbang Hasundutan      12.0      55.0   
1030  2019  Sumatera Utara             Kab. Samosir      25.0      10.5   
1031  2019  Sumatera Utara               Kota Medan      48.0       5.0   
1032  2019  Sumatera Utara    Kota Pematang Siantar       NaN       NaN   
1033  2019  Sumatera Utara       Kota Tebing Tinggi       NaN       NaN   

      Paper-cartons (%)  Plastic (%)  Metal(%)  Kain(%)  Karet- Kulit (%)  \
0                   NaN          NaN       NaN      NaN               NaN   
1                   2.0         18.0       2.0      5.0              17.0   
2                   5.0         26.0       5.0      3.0               2.0   
3                  10.7         18.2       3.1      2.6               1.8   
4                  19.3         35.7       4.0      1.0               3.0   
...                 ...          ...       ...      ...               ...   
1029               10.0         13.0       3.0      1.0               0.2   
1030               20.0         30.0       4.0      2.0               1.0   
1031               17.0         14.0       1.0      4.0               2.0   
1032                NaN          NaN       NaN      NaN               NaN   
1033                NaN          NaN       NaN      NaN               NaN   

      Kaca(%)  Lainnya(%)  
0         NaN         NaN  
1         4.0         NaN  
2         4.0         4.0  
3         2.0         6.5  
4         2.0        19.0  
...       ...         ...  
1029      0.8         5.0  
1030      0.5         7.0  
1031      1.0         8.0  
1032      NaN         NaN  
1033      NaN         NaN  

[1034 rows x 12 columns]

In [5]:
gdp = pd.read_csv('Data/GDP (BPS).csv')
melted_gdp = pd.melt(gdp, id_vars='Kabupaten/Kota', var_name='Year',value_name='GDP')
melted_gdp.drop(melted_gdp[melted_gdp['Year'] == '2016'].index, inplace=True)
melted_gdp
# only gdp dont have province in same col as kabkot

Kabupaten/Kota  Year          GDP
514        Simeulue  2017   1897038.43
515    Aceh Singkil  2017   2126448.17
516    Aceh Selatan  2017   4855005.68
517   Aceh Tenggara  2017   4245132.35
518      Aceh Timur  2017   9104763.67
...             ...   ...          ...
3593         Puncak  2022   1612809.92
3594        Dogiyai  2022   1445339.59
3595     Intan Jaya  2022   1362067.73
3596         Deiyai  2022   1574243.63
3597  Kota Jayapura  2022  35428724.03

[3084 rows x 3 columns]

In [6]:
gini = pd.read_csv('Data/Gini Index (BPS).csv')
gini.drop([0],inplace=True)
gini.drop(columns=['Unnamed: 7'],inplace=True)
gini.rename(columns=
            {"Unnamed: 1": "2017", 
             "Unnamed: 2": "2018", 
             "Unnamed: 3": "2019",
             "Unnamed: 4": "2020", 
             "Unnamed: 5": "2021", 
             "Unnamed: 6": "2022",},inplace=True)
melted_gini = pd.melt(gini, id_vars='Province/Regency/City', var_name='Year',value_name='Gini')
# there is a nan for 'BUTON TENGAH' in 2017 -- value is 0,2388. fill in
# Assuming 'BUTON TENGAH' is in the 'Regency/City' column and the NaN is in the 'Gini' column
# melted_gini.loc[(melted_gini['Province/Regency/City'] == 'BUTON TENGAH') & (melted_gini['Year'] == 2017), 'Gini'] = 0.2388
melted_gini

Province/Regency/City  Year   Gini
0                     ACEH  2017  0.329
1                 SIMEULUE  2017  0.307
2             ACEH SINGKIL  2017  0.321
3             ACEH SELATAN  2017  0.321
4            ACEH TENGGARA  2017  0.284
...                    ...   ...    ...
3289               Dogiyai  2022  0.295
3290            Intan Jaya  2022  0.299
3291                Deiyai  2022  0.150
3292         Kota Jayapura  2022  0.336
3293             INDONESIA  2022    NaN

[3294 rows x 3 columns]

In [7]:
hdi = pd.read_csv('Data/HDI Index (BPS).csv')
hdi.drop(columns=['2016','2023'],inplace=True)
melted_hdi = pd.melt(hdi, id_vars='Province/Regency/City', var_name='Year',value_name='HDI')
melted_hdi

Province/Regency/City  Year    HDI
0                     ACEH  2017  70.60
1                 Simeulue  2017  64.41
2             Aceh Singkil  2017  67.37
3             Aceh Selatan  2017  65.03
4            Aceh Tenggara  2017  68.09
...                    ...   ...    ...
3289               Dogiyai  2022  55.72
3290            Intan Jaya  2022  49.25
3291                Deiyai  2022  50.42
3292         Kota Jayapura  2022  80.61
3293             INDONESIA  2022  72.91

[3294 rows x 3 columns]

In [8]:
population = pd.read_csv('Data/Population (BPS).csv')
population.drop([0],inplace=True)
population.rename(columns=
            {"Unnamed: 1": "2017", 
             "Unnamed: 2": "2018", 
             "Unnamed: 3": "2019",
             "Unnamed: 4": "2020", 
             "Unnamed: 5": "2021", 
             "Unnamed: 6": "2022",},inplace=True)
population = population[['Province/Regency/City','2017','2018','2019','2020','2021','2022']]
population.drop([550,551,552,553],inplace=True)
melted_pop = pd.melt(population, id_vars='Province/Regency/City', var_name='Year',value_name='Population')
melted_pop

Province/Regency/City  Year Population
0                     ACEH  2017  5189466.0
1                 SIMEULUE  2017    91372.0
2             ACEH SINGKIL  2017    11949.0
3             ACEH SELATAN  2017   231893.0
4            ACEH TENGGARA  2017   208481.0
...                    ...   ...        ...
3289               Dogiyai  2022   119815.0
3290            Intan Jaya  2022   139236.0
3291                Deiyai  2022   102168.0
3292         Kota Jayapura  2022   410852.0
3293             INDONESIA  2022        NaN

[3294 rows x 3 columns]

In [9]:
area = pd.read_csv('Data/Land Area (Wiki).csv')
area

Kabupaten/kota  Luas wilayah (km2)[2]
0               Aceh               56834.77
1         Aceh Barat                2782.87
2    Aceh Barat Daya                1882.28
3         Aceh Besar                2891.48
4          Aceh Jaya                3872.35
..               ...                    ...
547       Raja Ampat                8034.44
548           Sorong                6544.23
549   Sorong Selatan                6594.31
550         Tambrauw               11529.18
551      Kota Sorong                 656.64

[552 rows x 2 columns]

# Merging indicators

## Let's first try just 1 indicator df with area 

### Step 1) Since the place names may be represented differently in different datasets (e.g. "Kota A" vs "A", "AB" vs "A B"), create a matching dictionary!

In [10]:
# Make all province, kabupaten, kota names in uppercase for both columns in the 2 datasets to prevent differences and failing to merge
area['Kabupaten/kota'] = area['Kabupaten/kota'].str.upper()
melted_pop['Province/Regency/City'] = melted_pop['Province/Regency/City'].str.upper()

In [11]:
import difflib

# Create a list of town names from each dataset
towns_melted = melted_pop['Province/Regency/City'].tolist()
towns_area = area['Kabupaten/kota'].tolist()

# Find close matches between town names
matches_melted_to_area = {}
matches_area_to_melted = {}

# Find matches from melted_pop to area
for town_melted in towns_melted:
    match = difflib.get_close_matches(town_melted, towns_area, n=1)
    if match:
        matches_melted_to_area[town_melted] = match[0]

# Find matches from area to melted_pop
for town_area in towns_area:
    match = difflib.get_close_matches(town_area, towns_melted, n=1)
    if match:
        matches_area_to_melted[town_area] = match[0]

# Print the matches
print("Matches from melted_pop to area:")
for town_melted, match_area in matches_melted_to_area.items():
    print(f"{town_melted} -> {match_area}")

Matches from melted_pop to area:
ACEH -> ACEH
SIMEULUE -> SIMEULUE
ACEH SINGKIL -> ACEH SINGKIL
ACEH SELATAN -> ACEH SELATAN
ACEH TENGGARA -> ACEH TENGGARA
ACEH TIMUR -> ACEH TIMUR
ACEH TENGAH -> ACEH TENGAH
ACEH BARAT -> ACEH BARAT
ACEH BESAR -> ACEH BESAR
PIDIE -> PIDIE
BIREUEN -> BIREUEN
ACEH UTARA -> ACEH UTARA
ACEH BARAT DAYA -> ACEH BARAT DAYA
GAYO LUES -> GAYO LUES
ACEH TAMIANG ->  ACEH TAMIANG
NAGAN RAYA -> NAGAN RAYA
ACEH JAYA -> ACEH JAYA
BENER MERIAH -> BENER MERIAH
PIDIE JAYA -> PIDIE JAYA
BANDA ACEH -> KOTA BANDA ACEH
SABANG -> SUBANG
LANGSA -> LANGKAT
LHOKSEUMAWE -> KOTA LHOKSEUMAWE
SUBULUSSALAM -> KOTA SUBULUSSALAM
SUMATERA UTARA -> SUMATERA UTARA
NIAS -> NIAS
MANDAILING NATAL -> MANDAILING NATAL
TAPANULI SELATAN -> TAPANULI SELATAN
TAPANULI TENGAH -> TAPANULI TENGAH
TAPANULI UTARA -> TAPANULI UTARA
LABUHAN BATU -> LABUHANBATU
ASAHAN -> ASAHAN
SIMALUNGUN -> SIMALUNGUN
DAIRI -> DAIRI
KARO -> KARO
DELI SERDANG -> DELI SERDANG
LANGKAT -> LANGKAT
NIAS SELATAN -> NIAS SELATAN

In [12]:
len(matches_melted_to_area)

545

#### Issue 1) Duplicate matches

In [13]:
# Create a list to store duplicate matches
duplicate_matches = []

# Check for duplicates
for key, value in matches_melted_to_area.items():
    if list(matches_melted_to_area.values()).count(value) > 1:
        duplicate_matches.append((key, value))

# Print duplicate matches
print("Duplicate Matches:")
for match in duplicate_matches:
    print(match)

Duplicate Matches:
('SABANG', 'SUBANG')
('LANGSA', 'LANGKAT')
('LANGKAT', 'LANGKAT')
('DKI JAKARTA', 'KOTA SURAKARTA')
('SUBANG', 'SUBANG')
('KOTA SURAKARTA', 'KOTA SURAKARTA')
('D I YOGYAKARTA', 'KOTA YOGYAKARTA')
('KOTA YOGYAKARTA', 'KOTA YOGYAKARTA')


Remember first value is the melted_pop value, aka the correct/reference one; in other words, it should be present. Out of these duplicate matches, 
- ('SABANG', 'SUBANG'): WRONG! should be SABANG, Kota Sabang
- ('LANGSA', 'LANGKAT'): WRONG! should be LANGSA, Kota Langsa
- ('LANGKAT', 'LANGKAT') -- correct
- ('DKI JAKARTA', 'KOTA SURAKARTA'): WRONG! should be DKI JAKARTA, Daerah Khusus Ibukota Jakarta
- ('SUBANG', 'SUBANG') -- correct
- ('KOTA SURAKARTA', 'KOTA SURAKARTA') -- correct 
- ('D I YOGYAKARTA', 'KOTA YOGYAKARTA'): WRONG! should be DI YOGYAKARTA, Daerah Istimewa Yogyakarta
- ('KOTA YOGYAKARTA', 'KOTA YOGYAKARTA') -- correct

In [14]:
# Update the matches_melted_to_area dictionary with corrected mappings
matches_melted_to_area.update({
    'SABANG': 'KOTA SABANG',
    'LANGSA': 'KOTA LANGSA',
    'DKI JAKARTA': 'DAERAH KHUSUS IBUKOTA JAKARTA',
    'D I YOGYAKARTA': 'DAERAH ISTIMEWA YOGYAKARTA'
})

# Check if there are any remaining duplicate matches
duplicate_matches = []
for key, value in matches_melted_to_area.items():
    if list(matches_melted_to_area.values()).count(value) > 1:
        duplicate_matches.append((key, value))

# Print the updated matches
print("Updated Matches:")
for match in matches_melted_to_area.items():
    print(match)

Updated Matches:
('ACEH', 'ACEH')
('SIMEULUE', 'SIMEULUE')
('ACEH SINGKIL', 'ACEH SINGKIL')
('ACEH SELATAN', 'ACEH SELATAN')
('ACEH TENGGARA', 'ACEH TENGGARA')
('ACEH TIMUR', 'ACEH TIMUR')
('ACEH TENGAH', 'ACEH TENGAH')
('ACEH BARAT', 'ACEH BARAT')
('ACEH BESAR', 'ACEH BESAR')
('PIDIE', 'PIDIE')
('BIREUEN', 'BIREUEN')
('ACEH UTARA', 'ACEH UTARA')
('ACEH BARAT DAYA', 'ACEH BARAT DAYA')
('GAYO LUES', 'GAYO LUES')
('ACEH TAMIANG', ' ACEH TAMIANG')
('NAGAN RAYA', 'NAGAN RAYA')
('ACEH JAYA', 'ACEH JAYA')
('BENER MERIAH', 'BENER MERIAH')
('PIDIE JAYA', 'PIDIE JAYA')
('BANDA ACEH', 'KOTA BANDA ACEH')
('SABANG', 'KOTA SABANG')
('LANGSA', 'KOTA LANGSA')
('LHOKSEUMAWE', 'KOTA LHOKSEUMAWE')
('SUBULUSSALAM', 'KOTA SUBULUSSALAM')
('SUMATERA UTARA', 'SUMATERA UTARA')
('NIAS', 'NIAS')
('MANDAILING NATAL', 'MANDAILING NATAL')
('TAPANULI SELATAN', 'TAPANULI SELATAN')
('TAPANULI TENGAH', 'TAPANULI TENGAH')
('TAPANULI UTARA', 'TAPANULI UTARA')
('LABUHAN BATU', 'LABUHANBATU')
('ASAHAN', 'ASAHAN')
('SIMALU

In [15]:
matches_set = set(matches_melted_to_area.values())
if len(matches_set) != len(matches_melted_to_area):
    print("Duplicate matches found!")
else:
    print("No duplicate matches.")

No duplicate matches.


In [16]:
len(matches_melted_to_area)

545

#### Issue 2) These towns are present in melted_pop but they couldn't find a matching entry in the area dataset.

In [17]:
# Get the unique town names from melted_pop
unique_towns_melted_pop = melted_pop['Province/Regency/City'].unique()

# Find unmatched towns
unmatched_towns = [town for town in unique_towns_melted_pop if town not in matches_melted_to_area]

# Print unmatched towns
print("Unmatched towns:")
for town in unmatched_towns:
    print(town)

Unmatched towns:
TOBA SAMOSIR / TOBA
KEP SERIBU
MAMUJU UTARA / PASANGKAYU
INDONESIA


In [18]:
# Define the mappings
new_mappings = {
    "TOBA SAMOSIR / TOBA": "TOBA",
    "KEP SERIBU": "ADMINISTRASI KEPULAUAN SERIBU",
    "MAMUJU UTARA / PASANGKAYU": "PASANGKAYU"
}

# Add new mappings
matches_melted_to_area.update(new_mappings)

# Delete specified entries
entries_to_delete = ["INDONESIA"]
for entry in entries_to_delete:
    if entry in matches_melted_to_area:
        del matches_melted_to_area[entry]

print("New mappings added and specified entries deleted successfully.")


New mappings added and specified entries deleted successfully.


In [19]:
len(matches_melted_to_area)

548

### Step 2) Now merge! After fixing all issues so have 548 unique matches, we expect 548 places x 6 years (2017-2022) = 3288 rows no NANs

In [20]:
# Swap keys and values in the dictionary
matches_melted_to_area_swapped = {v: k for k, v in matches_melted_to_area.items()}

# Use the swapped dictionary for mapping
area['Matched_town'] = area['Kabupaten/kota'].map(matches_melted_to_area_swapped)

# Merge datasets based on the matches from melted_pop to area
merged_data = pd.merge(melted_pop, area, how='left', left_on='Province/Regency/City', right_on='Matched_town')

# Assuming 'Indonesia' is in the column 'Matched_town', you can drop it like this
merged_data = merged_data.drop(merged_data[merged_data['Province/Regency/City'] == 'INDONESIA'].index)

# Drop redundant columns
merged_data.drop(columns=['Kabupaten/kota','Matched_town'], inplace=True)

# Display the new merged dataset
merged_data

Province/Regency/City  Year Population  Luas wilayah (km2)[2]
0                     ACEH  2017  5189466.0               56834.77
1                 SIMEULUE  2017    91372.0                1821.75
2             ACEH SINGKIL  2017    11949.0                1852.82
3             ACEH SELATAN  2017   231893.0                4175.38
4            ACEH TENGGARA  2017   208481.0                4179.12
...                    ...   ...        ...                    ...
3288                PUNCAK  2022   116279.0                8055.00
3289               DOGIYAI  2022   119815.0                4237.40
3290            INTAN JAYA  2022   139236.0                3922.02
3291                DEIYAI  2022   102168.0                 537.39
3292         KOTA JAYAPURA  2022   410852.0                 935.92

[3288 rows x 4 columns]

In [21]:
# Check for NaN values in the entire dataset
nan_values_exist = merged_data.isna().any().any()

if nan_values_exist:
    print("NaN values exist in the dataset.")
else:
    print("No NaN values found in the dataset.")

No NaN values found in the dataset.


In [22]:
merged_data[merged_data['Province/Regency/City']=='KEP SERIBU'] # ensuring its taking the values from melted_pop ds not area
# merged_data[merged_data['Kabupaten/kota']=='ADMINISTRASI KEPULAUAN SERIBU']

Province/Regency/City  Year Population  Luas wilayah (km2)[2]
165             KEP SERIBU  2017    23897.0                  10.18
714             KEP SERIBU  2018      24134                  10.18
1263            KEP SERIBU  2019      24295                  10.18
1812            KEP SERIBU  2020    27749.0                  10.18
2361            KEP SERIBU  2021    28240.0                  10.18
2910            KEP SERIBU  2022    28925.0                  10.18

## Now that we have merged 1 indicator df (population) w/ area, do the same for the rest of the indicators (GDP, HDI, Gini)

In [23]:
merged_data

Province/Regency/City  Year Population  Luas wilayah (km2)[2]
0                     ACEH  2017  5189466.0               56834.77
1                 SIMEULUE  2017    91372.0                1821.75
2             ACEH SINGKIL  2017    11949.0                1852.82
3             ACEH SELATAN  2017   231893.0                4175.38
4            ACEH TENGGARA  2017   208481.0                4179.12
...                    ...   ...        ...                    ...
3288                PUNCAK  2022   116279.0                8055.00
3289               DOGIYAI  2022   119815.0                4237.40
3290            INTAN JAYA  2022   139236.0                3922.02
3291                DEIYAI  2022   102168.0                 537.39
3292         KOTA JAYAPURA  2022   410852.0                 935.92

[3288 rows x 4 columns]

In [24]:
melted_gdp

Kabupaten/Kota  Year          GDP
514        Simeulue  2017   1897038.43
515    Aceh Singkil  2017   2126448.17
516    Aceh Selatan  2017   4855005.68
517   Aceh Tenggara  2017   4245132.35
518      Aceh Timur  2017   9104763.67
...             ...   ...          ...
3593         Puncak  2022   1612809.92
3594        Dogiyai  2022   1445339.59
3595     Intan Jaya  2022   1362067.73
3596         Deiyai  2022   1574243.63
3597  Kota Jayapura  2022  35428724.03

[3084 rows x 3 columns]

In [25]:
melted_gini['Province/Regency/City'] = melted_gini['Province/Regency/City'].str.upper()
melted_hdi['Province/Regency/City'] = melted_hdi['Province/Regency/City'].str.upper()
melted_gdp['Kabupaten/Kota'] = melted_gdp['Kabupaten/Kota'].str.upper()

In [26]:
# GINI
merged_data = pd.merge(merged_data, melted_gini, how='left', on=['Province/Regency/City', 'Year'], suffixes=('_merged', '_gini')) #suffixes to prevent duplicate error

# HDI
merged_data = pd.merge(merged_data, melted_hdi, how='left', on=['Province/Regency/City', 'Year'], suffixes=('_merged', '_hdi')) #suffixes to prevent duplicate error

# GDP
melted_gdp.rename(columns={'Kabupaten/Kota': 'Province/Regency/City'}, inplace=True)
merged_data = pd.merge(merged_data, melted_gdp, how='left', on=['Province/Regency/City', 'Year'], suffixes=('_merged', '_gdp')) #suffixes to prevent duplicate error

# Save this dataset to a new CSV file or use it further as needed
merged_data.to_csv("mergedindicators_withprovinces.csv", index=False)

# Display the final merged dataset
merged_data

Province/Regency/City  Year Population  Luas wilayah (km2)[2]   Gini  \
0                     ACEH  2017  5189466.0               56834.77  0.329   
1                 SIMEULUE  2017    91372.0                1821.75  0.307   
2             ACEH SINGKIL  2017    11949.0                1852.82  0.321   
3             ACEH SELATAN  2017   231893.0                4175.38  0.321   
4            ACEH TENGGARA  2017   208481.0                4179.12  0.284   
...                    ...   ...        ...                    ...    ...   
3283                PUNCAK  2022   116279.0                8055.00  0.252   
3284               DOGIYAI  2022   119815.0                4237.40  0.295   
3285            INTAN JAYA  2022   139236.0                3922.02  0.299   
3286                DEIYAI  2022   102168.0                 537.39  0.150   
3287         KOTA JAYAPURA  2022   410852.0                 935.92  0.336   

        HDI          GDP  
0     70.60          NaN  
1     64.41   1897038.43  
2     67.37   2126448.17  
3     65.03   4855005.68  
4     68.09   4245132.35  
...     ...          ...  
3283  43.87   1612809.92  
3284  55.72   1445339.59  
3285  49.25   1362067.73  
3286  50.42   1574243.63  
3287  80.61  35428724.03  

[3288 rows x 7 columns]

In [27]:
# since we just want the kabupaten/kota level, no need any province data -- some tables don't have it as well so to keep it simple, drop those rows
provinces_to_drop = {'ACEH', 'SUMATERA UTARA', 'SUMATERA BARAT', 'RIAU', 'KEPULAUAN RIAU', 'JAMBI', 'SUMATERA SELATAN', 'KEP BANGKA BELITUNG', 
                     'BENGKULU', 'LAMPUNG', 'DKI JAKARTA', 'JAWA BARAT', 'BANTEN', 'JAWA TENGAH', 'D I YOGYAKARTA', 'JAWA TIMUR', 'KALIMANTAN BARAT',
                     'KALIMANTAN TENGAH', 'KALIMANTAN SELATAN', 'KALIMANTAN TIMUR', 'KALIMANTAN UTARA', 'SULAWESI UTARA', 'GORONTALO', 
                     'SULAWESI TENGAH', 'SULAWESI SELATAN', 'SULAWESI BARAT', 'SULAWESI TENGGARA', 'BALI', 'NUSA TENGGARA BARAT', 
                     'NUSA TENGGARA TIMUR', 'MALUKU', 'MALUKU UTARA', 'PAPUA BARAT', 'PAPUA'}

merged_data.drop(merged_data[merged_data['Province/Regency/City'].isin(provinces_to_drop)].index, inplace=True)
merged_data.rename(columns={'Province/Regency/City': 'Regency/City'}, inplace=True)
merged_data

Regency/City  Year Population  Luas wilayah (km2)[2]   Gini    HDI  \
1          SIMEULUE  2017    91372.0                1821.75  0.307  64.41   
2      ACEH SINGKIL  2017    11949.0                1852.82  0.321  67.37   
3      ACEH SELATAN  2017   231893.0                4175.38  0.321  65.03   
4     ACEH TENGGARA  2017   208481.0                4179.12  0.284  68.09   
5        ACEH TIMUR  2017   419594.0                5409.41  0.265  66.32   
...             ...   ...        ...                    ...    ...    ...   
3283         PUNCAK  2022   116279.0                8055.00  0.252  43.87   
3284        DOGIYAI  2022   119815.0                4237.40  0.295  55.72   
3285     INTAN JAYA  2022   139236.0                3922.02  0.299  49.25   
3286         DEIYAI  2022   102168.0                 537.39  0.150  50.42   
3287  KOTA JAYAPURA  2022   410852.0                 935.92  0.336  80.61   

              GDP  
1      1897038.43  
2      2126448.17  
3      4855005.68  
4      4245132.35  
5      9104763.67  
...           ...  
3283   1612809.92  
3284   1445339.59  
3285   1362067.73  
3286   1574243.63  
3287  35428724.03  

[3084 rows x 7 columns]

In [28]:
len(provinces_to_drop)

34

## Yay! We have our merged dataset -- let's now do some quality control by checking for NANs, duplicates, any reordering of columns for user experience purposes

### A) Check for duplicates

In [29]:
# Check for duplicates
duplicates_mask = area.duplicated(subset=['Matched_town'], keep=False)
duplicated_rows = area[duplicates_mask].sort_values(by='Matched_town')

# Print duplicated rows
print(duplicated_rows)

       Kabupaten/kota  Luas wilayah (km2)[2] Matched_town
522     PAPUA SELATAN              127280.69          NaN
527      PAPUA TENGAH               62019.61          NaN
536  PAPUA PEGUNUNGAN               88416.64          NaN
545  PAPUA BARAT DAYA               38820.49          NaN


**no duplicates! yay!**

### B) Check for NANs

In [30]:
# Check for NaN values in the entire dataset
nan_values_exist = merged_data.isna().any().any()

if nan_values_exist:
    print("NaN values exist in the dataset.")
else:
    print("No NaN values found in the dataset.")

NaN values exist in the dataset.


#### i) in GDP column 

In [31]:
nan_mask_gdp = merged_data['GDP'].isna()
rows_with_nans_gdp = merged_data[nan_mask_gdp]
print("Rows with NaN values in GDP column:")
rows_with_nans_gdp.head()

Rows with NaN values in GDP column:


Regency/City  Year Population  Luas wilayah (km2)[2]   Gini  HDI  GDP
19    BANDA ACEH  2017   259913.0                  56.77  0.314  NaN  NaN
20        SABANG  2017    33978.0                 122.06  0.247  NaN  NaN
21        LANGSA  2017   171574.0                 224.24  0.342  NaN  NaN
22   LHOKSEUMAWE  2017    19898.0                 132.97  0.291  NaN  NaN
23  SUBULUSSALAM  2017    78725.0                1183.60  0.330  NaN  NaN

In [32]:
print(set(rows_with_nans_gdp['Regency/City']))
print(len(set(rows_with_nans_gdp['Regency/City'])))

{'LANGSA', 'KEP SERIBU', 'BANDA ACEH', 'LHOKSEUMAWE', 'MALUKU TENGGARA BARAT / KEPULAUAN TANIMBAR', 'KOTABARU', 'MAMUJU UTARA / PASANGKAYU', 'SUBULUSSALAM', 'SABANG', 'TOBA SAMOSIR / TOBA'}
10


**values not found, meaning this discrepancy is potentially due to names of places represented diff in GDP table compared to melted_pop table which is where the merged_data is getting its place names from -- for example, in GDP table it's 'toli-toli' but in merged it is 'tolitoli.' LETS FIX THIS!**

#### ii) in HDI column 

In [33]:
nan_mask_hdi = merged_data['HDI'].isna()
rows_with_nans_hdi = merged_data[nan_mask_hdi]
print("Rows with NaN values in HDI column:")
rows_with_nans_hdi.head()

Rows with NaN values in HDI column:


Regency/City  Year Population  Luas wilayah (km2)[2]   Gini  HDI  GDP
19    BANDA ACEH  2017   259913.0                  56.77  0.314  NaN  NaN
20        SABANG  2017    33978.0                 122.06  0.247  NaN  NaN
21        LANGSA  2017   171574.0                 224.24  0.342  NaN  NaN
22   LHOKSEUMAWE  2017    19898.0                 132.97  0.291  NaN  NaN
23  SUBULUSSALAM  2017    78725.0                1183.60  0.330  NaN  NaN

In [34]:
print(set(rows_with_nans_hdi['Regency/City']))
print(len(set(rows_with_nans_hdi['Regency/City'])))

{'LANGSA', 'KEP SERIBU', 'BANDA ACEH', 'LHOKSEUMAWE', 'SUBULUSSALAM', 'SABANG'}
6


#### iii) in Gini column 

In [35]:
nan_mask_gini = merged_data['Gini'].isna()
rows_with_nans_gini = merged_data[nan_mask_gini]
print("Rows with NaN values in Gini column:")
rows_with_nans_gini.head()

Rows with NaN values in Gini column:


Regency/City  Year Population  Luas wilayah (km2)[2]  Gini  HDI  GDP
165    KEP SERIBU  2017    23897.0                  10.18   NaN  NaN  NaN
713    KEP SERIBU  2018      24134                  10.18   NaN  NaN  NaN
1261   KEP SERIBU  2019      24295                  10.18   NaN  NaN  NaN
1809   KEP SERIBU  2020    27749.0                  10.18   NaN  NaN  NaN
2357   KEP SERIBU  2021    28240.0                  10.18   NaN  NaN  NaN

In [36]:
print(set(rows_with_nans_gini['Regency/City']))
print(len(set(rows_with_nans_gini['Regency/City'])))

{'KEP SERIBU'}
1


#### Let's just manually fix it so that no need to create matching dictionary, too much effort for just 10 places

In [37]:
places_with_nan_gdp = set(rows_with_nans_gdp['Regency/City'])
places_with_nan_hdi = set(rows_with_nans_hdi['Regency/City'])
places_with_nan_gini = set(rows_with_nans_gini['Regency/City'])

# Mapping dictionaries for each column
gdp_mapping = {
    'BANDA ACEH': 'KOTA BANDA ACEH',
    'KEP SERIBU': 'KEP. SERIBU',
    'LANGSA': 'KOTA LANGSA',
    'TOBA SAMOSIR / TOBA': 'TOBA SAMOSIR',
    'MALUKU TENGGARA BARAT / KEPULAUAN TANIMBAR': 'MALUKU TENGGARA BARAT',
    'SUBULUSSALAM': 'KOTA SUBULUSSALAM',
    'SABANG': 'KOTA SABANG',
    'LHOKSEUMAWE': 'KOTA LHOKSEUMAWE',
    'KOTABARU': 'KOTA BARU',
    'MAMUJU UTARA / PASANGKAYU': 'MAMUJU UTARA'
}

hdi_mapping = {
    'BANDA ACEH': 'KOTA BANDA ACEH',
    'KEP SERIBU': 'KEP. SERIBU',
    'LANGSA': 'KOTA LANGSA',
    'SUBULUSSALAM': 'KOTA SUBULUSSALAM',
    'SABANG': 'KOTA SABANG',
    'LHOKSEUMAWE': 'KOTA LHOKSEUMAWE',
}

gini_mapping = {
    'KEP SERIBU': 'KEP. SERIBU'
}

# Fill NaN values in GDP column
for place in places_with_nan_gdp:
    gdp_value = melted_gdp.loc[melted_gdp['Province/Regency/City'] == gdp_mapping[place], 'GDP'].values
    merged_data.loc[merged_data['Regency/City'] == place, 'GDP'] = gdp_value

# Fill NaN values in HDI column
for place in places_with_nan_hdi:
    hdi_value = melted_hdi.loc[melted_hdi['Province/Regency/City'] == hdi_mapping[place], 'HDI'].values
    merged_data.loc[merged_data['Regency/City'] == place, 'HDI'] = hdi_value

# Fill NaN values in Gini column
for place in places_with_nan_gini:
    gini_value = melted_gini.loc[melted_gini['Province/Regency/City'] == gini_mapping[place], 'Gini'].values
    merged_data.loc[merged_data['Regency/City'] == place, 'Gini'] = gini_value

# Display the updated merged dataset
merged_data

Regency/City  Year Population  Luas wilayah (km2)[2]   Gini    HDI  \
1          SIMEULUE  2017    91372.0                1821.75  0.307  64.41   
2      ACEH SINGKIL  2017    11949.0                1852.82  0.321  67.37   
3      ACEH SELATAN  2017   231893.0                4175.38  0.321  65.03   
4     ACEH TENGGARA  2017   208481.0                4179.12  0.284  68.09   
5        ACEH TIMUR  2017   419594.0                5409.41  0.265  66.32   
...             ...   ...        ...                    ...    ...    ...   
3283         PUNCAK  2022   116279.0                8055.00  0.252  43.87   
3284        DOGIYAI  2022   119815.0                4237.40  0.295  55.72   
3285     INTAN JAYA  2022   139236.0                3922.02  0.299  49.25   
3286         DEIYAI  2022   102168.0                 537.39  0.150  50.42   
3287  KOTA JAYAPURA  2022   410852.0                 935.92  0.336  80.61   

              GDP  
1      1897038.43  
2      2126448.17  
3      4855005.68  
4      4245132.35  
5      9104763.67  
...           ...  
3283   1612809.92  
3284   1445339.59  
3285   1362067.73  
3286   1574243.63  
3287  35428724.03  

[3084 rows x 7 columns]

In [38]:
# Check for NaN values in the entire dataset
nan_values_exist = merged_data.isna().any().any()

if nan_values_exist:
    print("NaN values exist in the dataset.")
else:
    print("No NaN values found in the dataset.")

No NaN values found in the dataset.


### C) Make dataset prettier

In [39]:
# # Reorder the columns to make it more intuitive order
desired_columns = ['Regency/City', 'Year', 'Luas wilayah (km2)[2]', 'Population', 'GDP', 'Gini', 'HDI']
merged_data = merged_data.reindex(columns=desired_columns)
# Rename any columns in indonesian still
merged_data.rename(columns={'Luas wilayah (km2)[2]': 'Land Area (km2)'}, inplace=True)
merged_data

Regency/City  Year  Land Area (km2) Population          GDP   Gini  \
1          SIMEULUE  2017          1821.75    91372.0   1897038.43  0.307   
2      ACEH SINGKIL  2017          1852.82    11949.0   2126448.17  0.321   
3      ACEH SELATAN  2017          4175.38   231893.0   4855005.68  0.321   
4     ACEH TENGGARA  2017          4179.12   208481.0   4245132.35  0.284   
5        ACEH TIMUR  2017          5409.41   419594.0   9104763.67  0.265   
...             ...   ...              ...        ...          ...    ...   
3283         PUNCAK  2022          8055.00   116279.0   1612809.92  0.252   
3284        DOGIYAI  2022          4237.40   119815.0   1445339.59  0.295   
3285     INTAN JAYA  2022          3922.02   139236.0   1362067.73  0.299   
3286         DEIYAI  2022           537.39   102168.0   1574243.63  0.150   
3287  KOTA JAYAPURA  2022           935.92   410852.0  35428724.03  0.336   

        HDI  
1     64.41  
2     67.37  
3     65.03  
4     68.09  
5     66.32  
...     ...  
3283  43.87  
3284  55.72  
3285  49.25  
3286  50.42  
3287  80.61  

[3084 rows x 7 columns]

# Merging plastic data

In [8]:
trash_amount.rename(columns={'Tahun':'Year',
                             'Provinsi':'Province',
                             'Kabupaten/Kota':'Regency/City',
                            }, inplace=True)
trash_amount

Year        Province             Regency/City Per day (ton)  \
0     2019            Aceh        Kab. Aceh Selatan         95.93   
1     2019            Aceh         Kab. Aceh Tengah        100.97   
2     2019            Aceh            Kab. Simeulue         43.96   
3     2019            Aceh           Kab. Gayo Lues         37.64   
4     2019            Aceh        Kab. Bener Meriah         64.89   
...    ...             ...                      ...           ...   
1029  2022  Sumatera Utara  Kab. Humbang Hasundutan         59.33   
1030  2022  Sumatera Utara               Kota Medan       1722.60   
1031  2022  Sumatera Utara    Kota Pematang Siantar        241.43   
1032  2022  Sumatera Utara       Kota Tebing Tinggi        124.63   
1033  2022  Sumatera Utara     Kota Padangsidimpuan        161.82   

     Per year (ton)  
0          35013.28  
1          36853.74  
2          16044.14  
3          13738.60  
4          23683.10  
...             ...  
1029       21653.73  
1030      628749.22  
1031       88121.44  
1032       45488.20  
1033       59065.98  

[1034 rows x 5 columns]

In [9]:
trash_comp.rename(columns={'Town':'Regency/City',
                            }, inplace=True)
plastic_comp = trash_comp[['Year','Province','Regency/City','Plastic (%)']]
plastic_comp

Year        Province             Regency/City  Plastic (%)
0     2022            Aceh        Kab. Aceh Selatan          NaN
1     2022            Aceh       Kab. Aceh Tenggara         18.0
2     2022            Aceh          Kab. Aceh Timur         26.0
3     2022            Aceh          Kab. Aceh Barat         18.2
4     2022            Aceh               Kab. Pidie         35.7
...    ...             ...                      ...          ...
1029  2019  Sumatera Utara  Kab. Humbang Hasundutan         13.0
1030  2019  Sumatera Utara             Kab. Samosir         30.0
1031  2019  Sumatera Utara               Kota Medan         14.0
1032  2019  Sumatera Utara    Kota Pematang Siantar          NaN
1033  2019  Sumatera Utara       Kota Tebing Tinggi          NaN

[1034 rows x 4 columns]

## There are NANs in plastic %! To impute, find mean or median of each province at that given year

In [10]:
plastic_comp[plastic_comp.isna().any(axis=1)]

Year        Province           Regency/City  Plastic (%)
0     2022            Aceh      Kab. Aceh Selatan          NaN
26    2020            Aceh         Kab. Aceh Jaya          NaN
29    2020            Aceh            Kota Sabang          NaN
31    2019            Aceh      Kab. Aceh Selatan          NaN
35    2019            Aceh      Kab. Bener Meriah          NaN
...    ...             ...                    ...          ...
1011  2021  Sumatera Utara            Kab. Asahan          NaN
1012  2021  Sumatera Utara              Kab. Toba          NaN
1023  2020  Sumatera Utara           Kota Sibolga          NaN
1032  2019  Sumatera Utara  Kota Pematang Siantar          NaN
1033  2019  Sumatera Utara     Kota Tebing Tinggi          NaN

[234 rows x 4 columns]

In [11]:
# Group data by 'Year' and 'Province' and calculate the mean using agg()
mean_by_province_year = plastic_comp.groupby(['Year', 'Province'])['Plastic (%)'].agg('mean')

def impute_nan(row):
    year = row['Year']
    province = row['Province']
    if pd.isna(row['Plastic (%)']):  # Check if NaN value
        # Get mean value for the province and year
        mean_value = mean_by_province_year.loc[(year, province)]
        return mean_value
    else:
        return row['Plastic (%)']  # Return original value if not NaN

# Apply the impute_nan function to fill NaN values in 'Plastic (%)' column
plastic_comp['Plastic (%)'] = plastic_comp.apply(impute_nan, axis=1)
plastic_comp

/var/folders/0j/n7xzy3zj28l9wpm__mh81z0m0000gp/T/ipykernel_95380/3108279913.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  plastic_comp['Plastic (%)'] = plastic_comp.apply(impute_nan, axis=1)


Year        Province             Regency/City  Plastic (%)
0     2022            Aceh        Kab. Aceh Selatan    21.060588
1     2022            Aceh       Kab. Aceh Tenggara    18.000000
2     2022            Aceh          Kab. Aceh Timur    26.000000
3     2022            Aceh          Kab. Aceh Barat    18.200000
4     2022            Aceh               Kab. Pidie    35.700000
...    ...             ...                      ...          ...
1029  2019  Sumatera Utara  Kab. Humbang Hasundutan    13.000000
1030  2019  Sumatera Utara             Kab. Samosir    30.000000
1031  2019  Sumatera Utara               Kota Medan    14.000000
1032  2019  Sumatera Utara    Kota Pematang Siantar    19.157143
1033  2019  Sumatera Utara       Kota Tebing Tinggi    19.157143

[1034 rows x 4 columns]

### Stil have some NANs remaining for the cities where no data for their province in that year, namely
1. *DKI Jakarta in 2021*
2. *Papus Barat Daya in 2022*
3. *Papus Barat Daya in 2021*
4. *Papua Pegunungan in 2021*
5. *Papua Tengah in 2022*
6. *Sulawesi Barat in 2022*
7. *Sulawesi Barat in 2021*
- **Maybe can try impute with the average of all years of that province.**

In [12]:
plastic_comp[plastic_comp.isna().any(axis=1)]

Year          Province               Regency/City  Plastic (%)
124  2021       DKI Jakarta      Kab. Adm. Kep. Seribu          NaN
125  2021       DKI Jakarta    Kota Adm. Jakarta Pusat          NaN
126  2021       DKI Jakarta    Kota Adm. Jakarta Utara          NaN
127  2021       DKI Jakarta    Kota Adm. Jakarta Barat          NaN
128  2021       DKI Jakarta  Kota Adm. Jakarta Selatan          NaN
129  2021       DKI Jakarta    Kota Adm. Jakarta Timur          NaN
709  2022  Papua Barat Daya                Kab. Sorong          NaN
710  2021  Papua Barat Daya        Kab. Sorong Selatan          NaN
711  2021  Papua Barat Daya              Kab. Tambrauw          NaN
715  2021  Papua Pegunungan            Kab. Lanny Jaya          NaN
722  2022      Papua Tengah                Kab. Mimika          NaN
747  2022    Sulawesi Barat                Kab. Mamuju          NaN
748  2022    Sulawesi Barat                Kab. Majene          NaN
749  2022    Sulawesi Barat         Kab. Mamuju Tengah          NaN
750  2021    Sulawesi Barat                Kab. Mamuju          NaN
751  2021    Sulawesi Barat                Kab. Majene          NaN
752  2021    Sulawesi Barat         Kab. Mamuju Tengah          NaN

In [13]:
# Group data by 'Province' and calculate the mean using agg()
mean_by_province = plastic_comp.groupby(['Province'])['Plastic (%)'].agg('mean')

def impute_nan_2(row):
    province = row['Province']
    if pd.isna(row['Plastic (%)']):  # Check if NaN value
        # Get mean value for the province and year
        mean_value = mean_by_province.loc[(province)]
        return mean_value
    else:
        return row['Plastic (%)']  # Return original value if not NaN

# Apply the impute_nan_2 function to fill NaN values in 'Plastic (%)' column
plastic_comp['Plastic (%)'] = plastic_comp.apply(impute_nan_2, axis=1)
plastic_comp

/var/folders/0j/n7xzy3zj28l9wpm__mh81z0m0000gp/T/ipykernel_95380/2570458513.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  plastic_comp['Plastic (%)'] = plastic_comp.apply(impute_nan_2, axis=1)


Year        Province             Regency/City  Plastic (%)
0     2022            Aceh        Kab. Aceh Selatan    21.060588
1     2022            Aceh       Kab. Aceh Tenggara    18.000000
2     2022            Aceh          Kab. Aceh Timur    26.000000
3     2022            Aceh          Kab. Aceh Barat    18.200000
4     2022            Aceh               Kab. Pidie    35.700000
...    ...             ...                      ...          ...
1029  2019  Sumatera Utara  Kab. Humbang Hasundutan    13.000000
1030  2019  Sumatera Utara             Kab. Samosir    30.000000
1031  2019  Sumatera Utara               Kota Medan    14.000000
1032  2019  Sumatera Utara    Kota Pematang Siantar    19.157143
1033  2019  Sumatera Utara       Kota Tebing Tinggi    19.157143

[1034 rows x 4 columns]

In [14]:
plastic_comp[plastic_comp.isna().any(axis=1)]

Year          Province     Regency/City  Plastic (%)
715  2021  Papua Pegunungan  Kab. Lanny Jaya          NaN

#### Still remaining NANs for cities where no data for their province across all years -- in this case only 1 year (2021) available and it is NAN so no reference available to impute

## Now that no NANs in plastic %, merge

In [15]:
# Merge the datasets on the relevant columns (Year, Province, and Kabupaten/Kota)
plastic_data = pd.merge(trash_amount, plastic_comp, on=['Year', 'Province', 'Regency/City'])

# Convert 'Per day (ton)' column to numeric
plastic_data['Per day (ton)'] = plastic_data['Per day (ton)'].str.replace(',', '').astype(float)

# Multiply 'Per day (ton)' by 'Plastic (%)' to get 'Amount of plastic per day'
plastic_data['Plastic per day (ton)'] = plastic_data['Per day (ton)'] * (plastic_data['Plastic (%)'] / 100)

# Multiply 'Per year (ton)' by 'Plastic (%)' to get 'Amount of plastic per year'
plastic_data['Per year (ton)'] = plastic_data['Per year (ton)'].str.replace(',', '').astype(float)
plastic_data['Plastic per year (ton)'] = plastic_data['Per year (ton)'] * (plastic_data['Plastic (%)'] / 100)

# Select only the relevant columns
plastic_data = plastic_data[['Year', 'Province', 'Regency/City', 'Plastic per day (ton)', 'Plastic per year (ton)']]

# Display the final dataset
plastic_data

Year        Province             Regency/City  Plastic per day (ton)  \
0     2019            Aceh        Kab. Aceh Selatan              17.468853   
1     2019            Aceh         Kab. Aceh Tengah              25.242500   
2     2019            Aceh            Kab. Simeulue              15.495900   
3     2019            Aceh           Kab. Gayo Lues               6.398800   
4     2019            Aceh        Kab. Bener Meriah              11.816469   
...    ...             ...                      ...                    ...   
1029  2022  Sumatera Utara  Kab. Humbang Hasundutan               7.712900   
1030  2022  Sumatera Utara               Kota Medan             243.683302   
1031  2022  Sumatera Utara    Kota Pematang Siantar              24.143000   
1032  2022  Sumatera Utara       Kota Tebing Tinggi              20.750895   
1033  2022  Sumatera Utara     Kota Padangsidimpuan              20.567322   

      Plastic per year (ton)  
0                6375.918288  
1                9213.435000  
2                5655.559350  
3                2335.562000  
4                4312.692510  
...                      ...  
1029             2814.984900  
1030            88944.436534  
1031             8812.144000  
1032             7573.785300  
1033             7507.286058  

[1034 rows x 5 columns]

In [16]:
plastic_data[plastic_data.isna().any(axis=1)]

Year          Province     Regency/City  Plastic per day (ton)  \
715  2021  Papua Pegunungan  Kab. Lanny Jaya                    NaN   

     Plastic per year (ton)  
715                     NaN

# Export! Moving on to use these datasets in EDA!

In [17]:
plastic_data

Year        Province             Regency/City  Plastic per day (ton)  \
0     2019            Aceh        Kab. Aceh Selatan              17.468853   
1     2019            Aceh         Kab. Aceh Tengah              25.242500   
2     2019            Aceh            Kab. Simeulue              15.495900   
3     2019            Aceh           Kab. Gayo Lues               6.398800   
4     2019            Aceh        Kab. Bener Meriah              11.816469   
...    ...             ...                      ...                    ...   
1029  2022  Sumatera Utara  Kab. Humbang Hasundutan               7.712900   
1030  2022  Sumatera Utara               Kota Medan             243.683302   
1031  2022  Sumatera Utara    Kota Pematang Siantar              24.143000   
1032  2022  Sumatera Utara       Kota Tebing Tinggi              20.750895   
1033  2022  Sumatera Utara     Kota Padangsidimpuan              20.567322   

      Plastic per year (ton)  
0                6375.918288  
1                9213.435000  
2                5655.559350  
3                2335.562000  
4                4312.692510  
...                      ...  
1029             2814.984900  
1030            88944.436534  
1031             8812.144000  
1032             7573.785300  
1033             7507.286058  

[1034 rows x 5 columns]

In [18]:
merged_data

NameError: name 'merged_data' is not defined

In [19]:
# Save this dataset to a new CSV file or use it further as needed
plastic_data.to_csv("plastic_2018_2022.csv", index=False)
merged_data.to_csv("5indicators_2017_2022.csv", index=False)